[Flaking Test] TestDeleteOrEvictWithDryRunServer/Eviction_enabled_with_server_dry-run

### Which jobs are flaking?

https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit&width=20

### Which tests are flaking?

`TestDeleteOrEvictWithDryRunServer/Eviction_enabled_with_server_dry-run`

### Since when has it been flaking?

[05/05/2026 11:49](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/138780/pull-kubernetes-unit/2051600063003627520)

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#ci-kubernetes-unit&width=20

### Reason for failure (if possible)

```
[sig-cli] k8s.io/kubectl/pkg: drain expand_less 	5s
{Failed  === RUN   TestDeleteOrEvictWithDryRunServer/Eviction_enabled_with_server_dry-run
==================
WARNING: DATA RACE
Write at 0x00c0003daa40 by goroutine 59:
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:194 +0x3a
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Previous write at 0x00c0003daa40 by goroutine 60:
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:194 +0x3a
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Goroutine 59 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38

Goroutine 60 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38
==================
==================
WARNING: DATA RACE
Read at 0x00c0003daa20 by goroutine 59:
  bytes.(*Buffer).tryGrowByReslice()
      /usr/local/go/src/bytes/buffer.go:134 +0x48
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:195 +0xb9
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Previous write at 0x00c0003daa20 by goroutine 60:
  bytes.(*Buffer).grow()
      /usr/local/go/src/bytes/buffer.go:155 +0x439
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:197 +0xc4
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Goroutine 59 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38

Goroutine 60 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38
==================
==================
WARNING: DATA RACE
Read at 0x00c000988400 by goroutine 59:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:392 +0x0
  bytes.growSlice()
      /usr/local/go/src/bytes/buffer.go:268 +0x16b
  bytes.(*Buffer).grow()
      /usr/local/go/src/bytes/buffer.go:169 +0x2b0
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:197 +0xc4
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Previous write at 0x00c000988400 by goroutine 60:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:392 +0x0
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:199 +0x118
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Goroutine 59 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38

Goroutine 60 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38
==================
==================
WARNING: DATA RACE
Write at 0x00c0003daa38 by goroutine 59:
  bytes.(*Buffer).grow()
      /usr/local/go/src/bytes/buffer.go:172 +0x3b1
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:197 +0xc4
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Previous read at 0x00c0003daa38 by goroutine 60:
  bytes.(*Buffer).Len()
      /usr/local/go/src/bytes/buffer.go:97 +0x64
  bytes.(*Buffer).grow()
      /usr/local/go/src/bytes/buffer.go:145 +0x3f
  bytes.(*Buffer).Write()
      /usr/local/go/src/bytes/buffer.go:197 +0xc4
  fmt.Fprintf()
      /usr/local/go/src/fmt/print.go:225 +0xaa
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:289 +0x29c
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods.gowrap1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:360 +0x13a

Goroutine 59 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38

Goroutine 60 (running) created at:
  k8s.io/kubectl/pkg/drain.(*Helper).evictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:283 +0x24d
  k8s.io/kubectl/pkg/drain.(*Helper).DeleteOrEvictPods()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain.go:264 +0x184
  k8s.io/kubectl/pkg/drain.TestDeleteOrEvictWithDryRunServer.func1()
      /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/kubectl/pkg/drain/drain_test.go:520 +0xfee
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:2036 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:2101 +0x38
==================
    testing.go:1712: race detected during execution of test
--- FAIL: TestDeleteOrEvictWithDryRunServer/Eviction_enabled_with_server_dry-run (0.00s)

=== RUN   TestDeleteOrEvictWithDryRunServer
--- FAIL: TestDeleteOrEvictWithDryRunServer (0.00s)
}
```

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig cli

--- a/staging/src/k8s.io/kubectl/pkg/drain/drain.go
+++ b/staging/src/k8s.io/kubectl/pkg/drain/drain.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"io"
 	"math"
+	"sync"
 	"time"
 
 	corev1 "k8s.io/api/core/v1"
@@ -268,8 +269,22 @@ func (d *Helper) DeleteOrEvictPods(pods []corev1.Pod) error {
 	return d.deletePods(pods, getPodFn)
 }
 
+// syncWriter serializes concurrent Write calls with a mutex.
+type syncWriter struct {
+	mu sync.Mutex
+	w  io.Writer
+}
+
+func (s *syncWriter) Write(p []byte) (n int, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.w.Write(p)
+}
+
 func (d *Helper) evictPods(pods []corev1.Pod, evictionGroupVersion schema.GroupVersion, getPodFn func(namespace, name string) (*corev1.Pod, error)) error {
 	returnCh := make(chan error, 1)
+	out := io.Writer(&syncWriter{w: d.Out})
+	errOut := io.Writer(&syncWriter{w: d.ErrOut})
 	// 0 timeout means infinite, we use MaxInt64 to represent it.
 	var globalTimeout time.Duration
 	if d.Timeout == 0 {
@@ -286,13 +301,13 @@ func (d *Helper) evictPods(pods []corev1.Pod, evictionGroupVersion schema.GroupV
 				switch d.DryRunStrategy {
 				case cmdutil.DryRunServer:
 					//nolint:errcheck
-					fmt.Fprintf(d.Out, "evicting pod %s/%s (server dry run)\n", activePod.Namespace, activePod.Name)
+					fmt.Fprintf(out, "evicting pod %s/%s (server dry run)\n", activePod.Namespace, activePod.Name)
 				default:
 					if d.OnPodDeletionOrEvictionStarted != nil {
 						d.OnPodDeletionOrEvictionStarted(&activePod, true)
 					}
 					//nolint:errcheck
-					fmt.Fprintf(d.Out, "evicting pod %s/%s\n", activePod.Namespace, activePod.Name)
+					fmt.Fprintf(out, "evicting pod %s/%s\n", activePod.Namespace, activePod.Name)
 				}
 				select {
 				case <-ctx.Done():
@@ -310,7 +325,7 @@ func (d *Helper) evictPods(pods []corev1.Pod, evictionGroupVersion schema.GroupV
 					return
 				} else if apierrors.IsTooManyRequests(err) {
 					//nolint:errcheck
-					fmt.Fprintf(d.ErrOut, "error when evicting pods/%q -n %q (will retry after %v): %v\n", activePod.Name, activePod.Namespace, d.EvictErrorRetryDelay, err)
+					fmt.Fprintf(errOut, "error when evicting pods/%q -n %q (will retry after %v): %v\n", activePod.Name, activePod.Namespace, d.EvictErrorRetryDelay, err)
 					time.Sleep(d.EvictErrorRetryDelay)
 				} else if !activePod.ObjectMeta.DeletionTimestamp.IsZero() && apierrors.IsForbidden(err) && apierrors.HasStatusCause(err, corev1.NamespaceTerminatingCause) {
 					// an eviction request in a deleting namespace will throw a forbidden error,
@@ -321,7 +336,7 @@ func (d *Helper) evictPods(pods []corev1.Pod, evictionGroupVersion schema.GroupV
 					// an eviction request in a deleting namespace will throw a forbidden error,
 					// if the pod is not marked deleted, we retry until it is.
 					//nolint:errcheck
-					fmt.Fprintf(d.ErrOut, "error when evicting pod %q from terminating namespace %q (will retry after %v): %v\n", activePod.Name, activePod.Namespace, d.EvictErrorRetryDelay, err)
+					fmt.Fprintf(errOut, "error when evicting pod %q from terminating namespace %q (will retry after %v): %v\n", activePod.Name, activePod.Namespace, d.EvictErrorRetryDelay, err)
 					time.Sleep(d.EvictErrorRetryDelay)
 				} else {
 					returnCh <- fmt.Errorf("error when evicting pods/%q -n %q: %v", activePod.Name, activePod.Namespace, err)
@@ -349,7 +364,7 @@ func (d *Helper) evictPods(pods []corev1.Pod, evictionGroupVersion schema.GroupV
 				onFinishFn:                      d.OnPodDeletionOrEvictionFinished,
 				globalTimeout:                   globalTimeout,
 				skipWaitForDeleteTimeoutSeconds: d.SkipWaitForDeleteTimeoutSeconds,
-				out:                             d.Out,
+				out:                             out,
 			}
 			_, err := waitForDelete(params)
 			if err == nil {

--- a/staging/src/k8s.io/kubectl/pkg/drain/drain.go
+++ b/staging/src/k8s.io/kubectl/pkg/drain/drain.go
@@ -380,13 +380,10 @@ func (d *Helper) evictPods(pods []corev1.Pod, evictionGroupVersion schema.GroupV
 
 	numPods := len(pods)
 	for doneCount < numPods {
-		select {
-		case err := <-returnCh:
-			doneCount++
-			if err != nil {
-				errors = append(errors, err)
-			}
+		if err := <-returnCh; err != nil {
+			errors = append(errors, err)
 		}
+		doneCount++
 	}
 
 	return utilerrors.NewAggregate(errors)

